runtime/sema/type_tags.go

```diff
@@ -369,6 +369,25 @@ var (
 // Methods
 
 func LeastCommonSuperType(types ...Type) Type {
+
+	superType := leastCommonSuperType(types...)
+	if superType == InvalidType {
+		return superType
+	}
+
+	// Do a sanity check to see if all types are
+	// actually subtypes of the inferred supertype.
+	for _, typ := range types {
+		if !IsSubType(typ, superType) {
+			return InvalidType
+		}
```

Comment:
you could add this sanity check at the end of `leastCommonSuperType`, and export that directly. the recursion could be accomplished with an internal closure or a fixedpoint combinator

Reply:
wait no, you can't have recursive nested functions can you... maybe rename `leastCommonSuperType` to make it more apparent that this is just a wrapper around a recursive worker fn

Reply:
yeah, `LeastCommonSuperType` is the public function (exported) to be used everywhere, and is just a wrapper around the `leastCommonSuperType` which is a local function. We have used this naming pattern elsewhere too to distinguish between a local function vs a global/exported function.